RFC: Crypted file store based on Fernet

frispete opened the pull request:
Hi,
since there's some movement in cryptography lately, here's an attempt to implement a Fernet-based file store for python-keyring. [Fernet Spec](https://github.com/fernet/spec/blob/master/Spec.md).

For the same reason, I added properties for scheme and version to FileBacked in order to have a chance to implement a meaningful _migrate procedure later on. The idea is to also compare the (encryption) scheme on load, if it exists in the file backend. I think, that the minor code refactorings are worth it. 

Regarding cryptography in this package:
Well, it's a pain due to the crypto state for python in general. keyczar is officially in [weak](https://github.com/google/keyczar#keyczar) state, pycrypto is stalled [(pycryptodome](https://github.com/Legrandin/pycryptodome) might be a better alternative), and then there is [cryptography](https://github.com/pyca/cryptography), that is *maintained* at least.

Regarding my fernet implementation:
I would have loved to use Scrypt (at least..) as KDF, but my distribution doesn't support openssl 1.1.0, yet, which is what the cryptography is based on. I resorted to using PBKDF2HMAC, which is acceptable for now, but since the KDF is the weakest spot, given the hash power, that is available today *worldwide*, that issue needs addressing. 

At the moment, I'm struggling how I signal decryption exceptions to the upper layers. cryptography raises cryptography.fernet.InvalidToken for non matching passwords in decrypt. Give an advice, please.

I'm considering another crypto file format, that is based on [Argon2](https://github.com/P-H-C/phc-winner-argon2), but that is adding a pretty experimental codebase as dependency... Now, you hopefully see my intentions behind scheme and version.

Please, share your thoughts, thanks.

Cheers,
Pete

jaraco commented:
I'm all for having better backends for supporting better crypto, especially for use cases where the primary, system-based backends aren't suitable, and your implementation here looks sound and well-reasoned.

I do want to caution against contributing to this package. I wish for this package to largely be a grab-bag of deprecated backends. If a fernet-based (or other improved crypto, file-system-based) backend would be useful to you, I suggest perhaps creating it as a separate library. In the same way that keyrings.alt exposes backends through the entry_points advertised in setup.py, any other third-party library can do the same.

If you were to implement such a library, then you would have intimate control over the implementation and maintenance, but also wouldn't be affiliated with this second-class package.

I recognize that such an implementation _might_ wish to build on the `file` and `file_base` modules in this package, and I would hate for your library to have to depend on keyrings.alt for that functionality, so if you want to go this route, it might be worth first factoring that base functionality back into keyring or into another support library.

I'd be happy to help with this effort, either guiding you in creating the new package or helping factor out the file support functionality.

Let me know what you think and how I can help.

frispete commented:
Hi Jason,

thank you for the encouraging reply. Much appreciated.

If you can imagine to let a file_base.py version similar to this patch enter the keyring project, that would be a great base. What troubles me from that move is testing. Do you have ideas on how to get this done consistently?

OTOH, file.py isn't appropriate for keyring IMHO (well, to be fair, all current encryption schemes, even those used by SecretStorage and kwallet*, are _weak_ at some point, but that weakness is outside the scope of keyring _itself_, of course). 

I dived into pycryptodome today, and I have to confess, that it's gotten a nice package (a state, that pycrypto itself never reached for me...). I'm _not_ a crypto expert, I'm just happen to have to deal with crypto *a* *lot* lately (again), and that I need to do it correctly now (unlike I did before a couple of times...). 

A strong file based key storage is *one* building block of it. I'm okay to create a new package for that, building on top of file_base.py, naming it keyring.something.

How would you like us to communicate? Not sure, if this is the best place. Being an old fashioned guy, I would prefer direct email exchange with you, as I have a few questions about implementation details concerning keyring.

jaraco commented:
> What troubles me from that move is testing. 

Given that there are no tests currently, I wouldn't be so offended if file_base had no tests after the move - as long as keyrings.alt continued to function after depending on the moved functionality, I would consider that suitable for acceptance. If you would like to develop tests for file_base within keyrings.alt prior to the move, that would also be an option I would be considering.

> `file.py` isn't appropriate for keyring.

Agreed. And it sounds like you're comfortable ignoring that behavior for the fernet keyring, so that's great.

> How would you like us to communicate?

I like to prefer conversations in open forums, so they can be referenced (hyperlinked) and left available for posterity. We can continue the conversation here if you have general questions, or you can open a PR in this project or the keyring project itself for the proposed changes. If you have just a question, I also have enabled [Gitter for keyring](https://gitter.im/jaraco/keyring). I may be able to respond quickly, but that only depends on when you catch me.

frispete commented:
Okay, thanks, that sounds like a plan. 

For an internal encryption scheme, I moved to pycryptodome today for the very reason of the availability of scrypt. Apart from this [issue](https://github.com/Legrandin/pycryptodome/issues/49), it is doing well. While at it, I also moved to AES-OCB for  [authenticated encryption](https://blog.cryptographyengineering.com/2012/05/19/how-to-choose-authenticated-encryption/). 

Apart from being the most powerful AE available today, actually I **like** the [license](http://web.cs.ucdavis.edu/~rogaway/ocb/license.htm) aspect of it: it explicitly excludes **military** use. Do you think, that this would disturb its utilisation in an keyring submodule?

jaraco commented:
Not at all. That's the beauty of a third party module is it can reach its target audience without impeding the license of the root library... though IANAL, so take that advice for what it's worth.

jaraco commented:
I'm pretty sure we're deferring this PR for others, but feel free to continue to comment or re-submit.
